src/iter.rs

```diff
@@ -1125,43 +767,468 @@ impl<'a> Chunks<'a> {
+                    return Some(text);
+                }
+            }
+        }
+    }
+}
+
+impl<'a> Iterator for Chunks<'a> {
+    type Item = &'a str;
+
+    /// Advances the iterator forward and returns the next value.
+    ///
+    /// Runs in amortized O(1) time and worst-case O(log N) time.
+    #[inline(always)]
+    fn next(&mut self) -> Option<&'a str> {
+        if !self.is_reversed {
+            self.next_impl()
+        } else {
+            self.prev_impl()
+        }
+    }
+}
+
+//==========================================================
+
+/// An iterator over a `Rope`'s contiguous `str` chunks.
+///
+/// Internally, each `Rope` stores text as a segemented collection of utf8
+/// strings. This iterator iterates over those segments, returning a
+/// `&str` slice for each one.  It is useful for situations such as:
+///
+/// - Writing a rope's utf8 text data to disk (but see
+///   [`write_to()`](crate::rope::Rope::write_to) for a convenience function that does this
+///   for casual use-cases).
+/// - Streaming a rope's text data somewhere.
+/// - Saving a rope to a non-utf8 encoding, doing the encoding conversion
+///   incrementally as you go.
+/// - Writing custom iterators over a rope's text data.
+///
+/// There are precisely two guarantees about the yielded chunks:
+///
+/// - All non-empty chunks are yielded, and they are yielded in order.
+/// - CRLF pairs are never split across chunks.
+///
+/// There are no guarantees about the size of yielded chunks, and except for
+/// CRLF pairs and being valid `str` slices there are no guarantees about
+/// where the chunks are split.  For example, they may be zero-sized, they
+/// don't necessarily align with line breaks, etc.
+#[derive(Debug, Clone)]
+pub struct Lines<'a> {
+    iter: LinesEnum<'a>,
+    is_reversed: bool,
+    text: &'a str,
+    byte_idx: usize,
+    at_end: bool,
+    line_idx: usize,
+    total_lines: usize,
+}
+
+#[derive(Debug, Clone)]
+enum LinesEnum<'a> {
+    Full {
+        node_stack: Vec<(&'a Arc<Node>, usize)>, // (node ref, index of current child)
+        leaf_byte_idx: u32,
+        total_bytes: usize,
+    },
+    Light,
+}
+
+impl<'a> Lines<'a> {
+    pub(crate) fn new(node: &Arc<Node>) -> Lines {
+        let info = node.text_info();
+        Lines::new_with_range(
+            node,
+            (0, info.bytes as usize),
+            (0, info.line_breaks as usize + 1),
+        )
+    }
+
+    pub(crate) fn new_with_range(
+        node: &Arc<Node>,
+        byte_idx_range: (usize, usize),
+        line_idx_range: (usize, usize),
+    ) -> Lines {
+        Lines::new_with_range_at_impl::<false>(node, 0, byte_idx_range, line_idx_range)
+    }
+
+    pub(crate) fn new_with_range_at(
+        node: &Arc<Node>,
+        at_line: usize,
+        byte_idx_range: (usize, usize),
+        line_idx_range: (usize, usize),
+    ) -> Lines {
+        Lines::new_with_range_at_impl::<true>(node, at_line, byte_idx_range, line_idx_range)
+    }
```

Comment:
I may be misunderstanding the code here, but it looks to me like `new_with_range()` and `new_with_range_at()` are inconsistent with each other.  The following two calls should be identical:

```rust
Lines::new_with_range(node, byte_range, line_range);
Lines::new_with_range_at(node, 0, byte_range, line_range);
```

But it looks like they're different.  Specifically, the former calls `Lines::new_with_range_at_impl()` with a `false` const generic parameter, and the latter with `true`.

I'm also a little nervous about that const generic parameter existing at all: I would expect it to either always be true or always be false.  I haven't looked at the code closely enough yet, though, so I could certainly be wrong about that.  But it smells a little to me.

Reply:
Ok this is definitly subtle but its the most elegant way to implement it I could come up with. I will try to explain:

First a small nitpick about your comment so it doesn't get confusing later: 
The two iterators you posted are not always equivalent. They are only equivalent when you are iterating a full `Rope` (or a `RopeSlice` that happens to start within the first line of its top `Node`). However these two iterators are always equal if you use `line_range.0` instead of `0`:

``` rust
Lines::new_with_range(node, byte_range, line_range) 
Lines::new_with_range_at(node, line_range.0, byte_range, line_range)
```

Its trivial to see that my code fulfill that criteria, because if you look into the `new_with_range_at_impl` function it actually special cases `line == line_range.0` to always return `Lines::new_with_range(node, byte_range, line_range)` right at the start of the function.
The reason this special case is required is because the first line might start before the start of the `RopeSlice` (so before `byte_range.0`). That means the line start is potentially in a different chunk then the position you are looking for. To find the correct chunk you need to perform a tree search for the byte position `byte_range.0` instead of the line `line_range.0`.

So really the const generic indicates:
* `false` => start the iterator at the first line
* `true` => start the iterator at the specified line (with a dynamic check if this line is the first line to switch to the other implementation in that case)

This special case might seem odd but I will try to explain my reasoning a bit more:
A similar problem occurs at the end of the `line_range` (so `line >= line_range.1`). I handle this special case as follows:
``` rust
let mut res = Lines::new_with_range_at(node, line_range.1 - 1, byte_range, line_range);
res.next();
res
```
This reuses the logic I already have inside the line iterator to correctly handle the end of the last line and therefore doesn't require yet another specical cased implementation of the `new` function.

You might considered doing the inverse for the start of the line:
``` rust
let mut res = Lines::new_with_range_at(node, line_range.0 + 1, byte_range, line_range);
res.prev();
res
```

However this doesn't work when there is only a single line (because then these two cases just endlessly recuse into each other) so you need to handle one of them as a special case. 
Its admittedly somewhat of arbitrary which one of those two you special case but there are a couple reasons I special cased the first line instead of the last line:
* the implementation is slightly easier/more natural
* the special case is slightly faster (because it does not involve stepping the iterator an additional time) and constructing an iterator at the first line should be way more common (just `lines`) then constructing an iterator at the last line

I think its a bit easier to think about `new_with_range` and `new_with_range_at` as two seperate functions (where `new_with_range_at` will call `new_with_range` for `line == line_range.0`).
These two functions are just very similar so I used the const generic here to share most of the implementation.

Reply:
> First a small nitpick about your comment so it doesn't get confusing later:

Ah, right.  Thanks for pointing that out!

> The reason this special case is required is because the first line might start before the start of the RopeSlice (so before byte_range.0). That means the line start is potentially in a different chunk then the position you are looking for. To find the correct chunk you need to perform a tree search for the byte position byte_range.0 instead of the line line_range.0.

I'll keep an eye out for this when I take a closer look at the code.  But at first blush, I think I would have implemented the tree traversal code to account for both the line *and* byte starts, diving into the the furthest-forward of the two as it goes.  Then (unless I'm still missing something, which is possible) there wouldn't need to be any special cases.

> A similar problem occurs at the end of the line_range (so line >= line_range.1). I handle this special case as follows:

I guess part of what's throwing me off here is that it's not clear to me why these cases aren't handled in the traversal/stack-building code itself.  It *feels* to me like there's a simpler implementation struggling to emerge here, but we haven't quite hit it yet.

Partly this is from experiences with my own code in the past: special cases that look a lot like this have very often indicated that my perspective on the problem is a little off.

But to be clear, that doesn't mean the code is incorrect.  Not at all.  Just that a more straightforward and also correct alternative implementation may be possible.

Having said all of that, I don't mean to block this PR on stuff like that.  As long as it's correct and I can follow it, let's get this merged.  Especially considering that there's already a lot of "too complex" code in Ropey already, which is my own doing.  Ha ha. :-)  I can always revisit the code and try to simplify it myself in the future.

Reply:
I thaught about your suggestion regarding doing the tree sesach for lines/bytes all at once but my intuition is that while its theoretically possible it would involve negative indexing (because you always have to keep track of your position relatively to the current node) and a bunch of extra conditions to handle that correctly so the sketch I have in my mind right now is actually more complex then the current implementation.

I can experiment a bit with it (the ideas in my mind are often wrong :D) and if I find s better solution in the future I might do a follow-up PR if you can live with the current implementation for now. This function took a while to to get right (edgecases like a rope slice split inside a CRLF must be considered correctlly) and so overall improving this would take a while